src/core/allow-bind.c

```diff
@@ -0,0 +1,164 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#if HAVE_LIBBPF
+#include <bpf/bpf.h>
+#endif
+
+#include "allow-bind.h"
+#include "fd-util.h"
+#include "unit.h"
+
+#if BPF_FRAMEWORK
+/* libbpf, clang and llc compile time dependencies are satisfied */
+#include "bpf/allow_bind/allow-bind-skel.h"
+
+static struct allow_bind_bpf *allow_bind_bpf_free(struct allow_bind_bpf *obj) {
+        if (obj && libbpf_get_error(obj) == 0)
+                (void) allow_bind_bpf__destroy(obj);
+
+        return NULL;
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(struct allow_bind_bpf *, allow_bind_bpf_free);
+
+static int can_link_bpf_program(struct bpf_program *prog) {
+        _cleanup_(bpf_link_freep) struct bpf_link *link = NULL;
+        int err;
+
+        assert(prog);
+
+        link = bpf_program__attach_cgroup(prog, -1);
+        if (!link)
+                return -ENOMEM;
+
+        err = libbpf_get_error(link);
+        if (err != -EINVAL && err != -EBADF)
+                return err;
+
+        /* EBADF indicates that bpf_link is supported by kernel. */
+        return err == -EBADF;
+}
+
+static int prepare_allow_bind_bpf(Unit *u, const Set *ports_v6, const Set *ports_v4, struct allow_bind_bpf **ret_obj) {
+        _cleanup_(allow_bind_bpf_freep) struct allow_bind_bpf *obj = 0;
+        int map_fd, r = 0;
+        uint8_t dummy;
+        void *port;
+
+        obj = allow_bind_bpf__open();
+        if (!obj)
+                return log_unit_error_errno(u, r, "Failed to open BPF object");
+
+        r = bpf_map__resize(obj->maps.ports_v6, MAX(set_size(ports_v6), 1u));
+        if (r)
+                return log_unit_error_errno(u, r,
+                                "Failed to resize BPF map '%s': %m",
+                                bpf_map__name(obj->maps.ports_v6));
+
+        r = bpf_map__resize(obj->maps.ports_v4, MAX(set_size(ports_v4), 1u));
+        if (r)
+                return log_unit_error_errno(u, r,
+                                "Failed to resize BPF map '%s': %m",
+                                bpf_map__name(obj->maps.ports_v4));
+
+        r = allow_bind_bpf__load(obj);
+        if (r)
+                return log_unit_error_errno(u, r, "Failed to load BPF object");
+
+        map_fd = bpf_map__fd(obj->maps.ports_v6);
+        if (map_fd < 0)
+                return log_unit_error_errno(u, map_fd, "Failed to get BPF map fd");
+
+        SET_FOREACH(port, ports_v6)
+                if (bpf_map_update_elem(map_fd, &port, &dummy, BPF_ANY))
+                        return log_unit_error_errno(u, -1, "Failed to update BPF map");
```

Comment:
bpf_map_update_elem returns no valid error code?

if so, don't return -1, but make up a proper one, and indicate so with SYNTHETIC_ERRNO. i.e. `SYNTHETIC_ERRNO(ENOMEM)` might be right here

Reply:
but actually i think the function *does* return an error code, no?